### PR TITLE
Allow the receiver to request a preferred resolution

### PR DESCRIPTION
**Pull request by DemiMarie**
The sender tries to find a resolution that is close to the preferred one of the receiver.

Fixes https://github.com/QubesOS/qubes-issues/issues/8424

### CODE REVIEW

File: sender/webcam.py [hunk 1]
```diff
@@ -15,8 +15,43 @@
...
 
-    def __init__(self):
-        self.main(self)
+    untrusted_requested_width: int
+    untrusted_requested_height: int
+    untrusted_requested_fps: int
+
+    def __init__(self, *, untrusted_arg: str):
+        if untrusted_arg:
+            untrusted_arg_bytes = untrusted_arg.encode('ascii', 'strict')
```
Comment (marmarek) [inline sender/webcam.py:24]: You've got `str` already, there is no point in converting back to bytes.

**Review by marmarek — CHANGES_REQUESTED** [review]
Please add also setting the argument in `qubes-video-companion` script. Something like `qubes-video-companion --resolution w+h+f webcam` , maybe also with short option `-r` and maybe for convenience also with replacing `x` and `@` with `+`, so one can write `800x600@30` (but please, keep it simple, no complex regexes).

**Comment by marmarek** [comment]
> Please add also setting the argument in `qubes-video-companion` script. Something like `qubes-video-companion --resolution w+h+f webcam` , maybe also with short option `-r` and maybe for convenience also with replacing `x` and `@` with `+`, so one can write `800x600@30` (but please, keep it simple, no complex regexes).

Ping?

**Comment by DemiMarie** [comment]
> > Please add also setting the argument in `qubes-video-companion` script. Something like `qubes-video-companion --resolution w+h+f webcam` , maybe also with short option `-r` and maybe for convenience also with replacing `x` and `@` with `+`, so one can write `800x600@30` (but please, keep it simple, no complex regexes).
> 
> Ping?

I just pushed a version that implements it.  I’ll start testing it, but I wanted to get this out there first.

File: receiver/qubes-video-companion [hunk 2]
```diff
@@ -4,18 +4,44 @@
...
+set -u -e
 
 set -E # Enable function inheritance of traps
 trap exit ERR
+unset GETOPT_COMPATIBLE
+unsupported_getopt () {
+    echo "Your getopt(1) doesn't support quoting" >&2
+    exit 1
+}
+if getopt -T > /dev/null; then unsupported_getopt; elif [ "$?" -ne 4 ]; then unsupported_getopt; fi
```
Comment (marmarek) [inline receiver/qubes-video-companion:16]: The intended use does not require any spaces in arguments, so I think this check is overzealous (I know **incorrect** use with no quoting would result in a weird error, but I don't think it's really relevant).

File: receiver/qubes-video-companion [hunk 3]
```diff
@@ -4,18 +4,44 @@
...
+name=${0##*/}
 
 usage() {
-    echo "Usage: qubes-video-companion webcam|screenshare [destination qube]"
-} >&2
+    printf '%s: Usage: qubes-video-companion webcam|screenshare [destination qube]\n' "$name"
```
Comment (marmarek) [inline receiver/qubes-video-companion:20]: You did changed this line in the recent push, yet you did not added the new option...

**Review by marmarek — CHANGES_REQUESTED** [review]

**Comment by marmarek** [comment]
Shellcheck complains

**Comment by DemiMarie** [comment]
The shell script wrapper was not passing the argument to the Python script.  Now fixed.

**Comment by DemiMarie** [comment]
done

**Review by marmarek — APPROVED** [review]

**Pull request merged**